---
title: "scRNAseq Analysis - Integration"
output: html_notebook
author: Talia Wiggen
---
# Description
This notebook is adapted from: https://hbctraining.github.io/scRNA-seq_online/schedule/links-to-lessons.html

The objective of this notebook is to perform integration of cells across conditions to identify cells that are similar to each other.

Goals:
 - To align same cell types across conditions.

Challenges:
 - Aligning cells of similar cell types so that we do not have clustering downstream due to differences between samples, conditions, modalities, or batches

Recommendations:
 - Go through the analysis without integration first to determine whether integration is necessary


# Setup Environment
```{r}
# Load libraries
library(Seurat)
library(tidyverse)
library(cowplot)
library(RCurl)
```

# Load the data
```{r}
split_seurat <- readRDS("C:\\Users\\Hunter Lab\\Desktop\\Talia\\Fuso_SC_2021\\data\\seurat\\split_seurat.rds")
```


# Integrate or align samples across conditions using shared highly variable genes
If cells cluster by sample, condition, batch, dataset, modality, this integration step can greatly improve the clustering and the downstream analyses.

To integrate, we will use the shared highly variable genes (identified using SCTransform) from each group, then, we will “integrate” or “harmonize” the groups to overlay cells that are similar or have a “common set of biological features” between groups.  

Integration is a powerful method that uses these shared sources of greatest variation to identify shared subpopulations across conditions or datasets [Stuart and Bulter et al. (2018)]. The goal of integration is to ensure that the cell types of one condition/dataset align with the same celltypes of the other conditions/datasets (e.g. control macrophages align with stimulated macrophages).

Steps of integration:
1. Perform canonical correlation analysis (CCA):
CCA identifies shared sources of variation between the conditions/groups. It is a form of PCA, in that it identifies the greatest sources of variation in the data, but only if it is shared or conserved across the conditions/groups (using the 3000 most variant genes from each sample). This step roughly aligns the cells using the greatest shared sources of variation.

2. Identify anchors or mutual nearest neighbors (MNNs) across datasets:
For each cell in one condition: The cell’s closest neighbor in the other condition is identified based on gene expression values. The reciprocal analysis is performed, and if the two cells are closest neighbors in both directions, then those cells will be marked as anchors to ‘anchor’ the two datasets together.
 
3. Filter anchors to remove incorrect anchors:
Assess the similarity between anchor pairs by the overlap in their local neighborhoods (incorrect anchors will have low scores) - do the adjacent cells have ‘best buddies’ that are adjacent to each other?

4. Integrate the conditions/datasets:
Use anchors and corresponding scores to transform the cell expression values, allowing for the integration of the conditions/datasets.



First, we need to specify that we want to use all of the 3000 most variable genes identified by SCTransform for the integration. By default, this function only selects the top 2000 genes.

```{r}
# Select the most variable features to use for integration
integ_features <- SelectIntegrationFeatures(object.list = split_seurat, 
                                            nfeatures = 3000) 
```

Now, we need to prepare the SCTransform object for integration.
```{r}
# Prepare the SCT list object for integration
split_seurat <- PrepSCTIntegration(object.list = split_seurat, 
                                   anchor.features = integ_features)
```

Now, we are going to perform CCA, find the anchors and filter incorrect anchors.
```{r}
# Find anchors - can take a while to run
integ_anchors <- FindIntegrationAnchors(object.list = split_seurat, 
                                        normalization.method = "SCT", 
                                        anchor.features = integ_features)
```

Finally, we can integrate across conditions.
```{r}
# Integrate across conditions
seurat_integrated <- IntegrateData(anchorset = integ_anchors, 
                                   normalization.method = "SCT")
```

# UMAP Visualization
After integration, to visualize the integrated data we can use dimensionality reduction techniques, such as PCA and Uniform Manifold Approximation and Projection (UMAP). While PCA will determine all PCs, we can only plot two at a time. In contrast, UMAP will take the information from any number of top PCs to arrange the cells in this multidimensional space. It will take those distances in multidimensional space and plot them in two dimensions working to preserve local and global structure. In this way, the distances between cells represent similarity in expression. 

To generate these visualizations we need to first run PCA and UMAP methods. Let’s start with PCA.
```{r}
# Run PCA
seurat_integrated <- RunPCA(object = seurat_integrated)

# Plot PCA
## split by sample
PCAPlot(seurat_integrated,
        split.by = "sample")

## split by condition
PCAPlot(seurat_integrated,
        split.by = "condition") 
```

Now, lets visualize with UMAP
```{r}
# Run UMAP
seurat_integrated <- RunUMAP(seurat_integrated, 
                             dims = 1:40,
			     reduction = "pca")

# Plot UMAP                             
DimPlot(seurat_integrated) 

# Plot UMAP split by sample
DimPlot(seurat_integrated,
        split.by = "sample")  

# Plot UMAP split by condition
DimPlot(seurat_integrated,
        split.by = "condition")
```


```{r}
# Save integrated seurat object
saveRDS(seurat_integrated, "C:\\Users\\Hunter Lab\\Desktop\\Talia\\Fuso_SC_2021\\data\\seurat\\integrated_seurat.rds")
```

