---
title: "scRNAseq Analysis - Quality Control and Filtering"
output: html_notebook
author: Talia Wiggen
---
# Description
This notebook is adapted from: https://hbctraining.github.io/scRNA-seq_online/schedule/links-to-lessons.html

The goal of this notebook is to:
1. Construct quality control metrics and visually evaluate the quality of the data
2. Apply appropriate filters to remove low quality cells

# Setup Environment
```{r}
# Load libraries
library(SingleCellExperiment)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)
```
# Load the data

For each individual sample you will have the following three files from CellRanger output:
1. barcodes.tsv - a file with the cell IDs, representing all cells quantified 
2. features.tsv - a file with the gene IDs, representing all genes quantified
3. matrix.mtx - a matrix of counts per gene for every cell

While Cell Ranger performs filtering on the expression counts (see note below), we wish to perform our own QC and filtering because we want to account for the biology of our experiment/biological system. Given this we are only interested in the raw_feature_bc_matrix folder in the Cell Ranger output.

```{r}
# Create a Seurat object for each sample
for (file in c("ana_1", "ana_2", "ana_3", "ana_4", "ana_5", "ana_6", "fuso_1", "fuso_2", "fuso_3")){ #CHANGE me to your sample folders that contain the raw barcodes, features, and matrix files
        seurat_data <- Read10X(data.dir = paste0("../data/cellranger/raw_feature_bc_matrices/", file)) #CHANGE me to your filepath to the parent folder of the sample folder from the last line
        seurat_obj <- CreateSeuratObject(counts = seurat_data, 
                                         min.features = 100, 
                                         project = file)
        assign(file, seurat_obj)
}
```
NOTE: The min.features argument specifies the minimum number of genes that need to be detected per cell. This argument will filter out poor quality cells that likely just have random barcodes encapsulated without any cell present. Usually, cells with less than 100 genes detected are not considered for analysis.

```{r}
# Check the metadata in the new Seurat objects
head(ana_1@meta.data)
head(ana_2@meta.data)
head(ana_3@meta.data)
head(ana_4@meta.data)
head(ana_5@meta.data)
head(ana_6@meta.data)
head(fuso_1@meta.data)
head(fuso_2@meta.data)
head(fuso_3@meta.data)
```
Seurat automatically creates some metadata for each of the cells when you use the Read10X() function to read in data. This information is stored in the meta.data slot within the Seurat object.
 - orig.ident: this often contains the sample identity if known, but will default to “SeuratProject”
 - nCount_RNA: number of UMIs per cell
 - nFeature_RNA: number of genes detected per cell



Next, we need to merge these objects together into a single Seurat object. This will make it easier to run the QC steps for both sample groups together and enable us to easily compare the data quality for all the samples.

We can use the merge() function from the Seurat package to do this:
```{r}
# Create a merged Seurat object
merged_seurat <- merge(x = ana_1, 
                       y = c(ana_2, ana_3, ana_4, ana_5, ana_6, fuso_1, fuso_2, fuso_3), 
                       add.cell.id = c("ana1", "ana2", "ana3", "ana4", "ana5", "ana6", "fuso1", "fuso2", "fuso3")) #we add a sample-specific prefix to each of our cell IDs 
                       
```

```{r}
#check sample prefix for merged object
unique(sapply(X = strsplit(colnames(merged_seurat), split = "_"), FUN = "[", 1))
table(merged_seurat$orig.ident)
```

# Quality Control
Goals:
 - To filter the data to only include true cells that are of high quality, so that when we cluster our cells it is easier to identify distinct cell type populations
 - To identify any failed samples and either try to salvage the data or remove from analysis, in addition to, trying to understand why the sample failed

Challenges:
 - Delineating cells that are poor quality from less complex cells
 - Choosing appropriate thresholds for filtering, so as to keep high quality cells without removing biologically relevant cell types
 
Recommendations:
 - Have a good idea of your expectations for the cell types to be present prior to performing the QC. For instance, do you expect to have low complexity cells or cells with higher levels of mitochondrial expression in your sample? If so, then we need to account for this biology when assessing the quality of our data.

## Generating Quality Metrics
As mentioned earlier, Seurat object contains metadata with the orig.ident, nCount_RNA, and nFeature_RNA.In order to create the appropriate plots for the quality control analysis, we need to calculate some additional metrics. These include:
 - number of genes detected per UMI: this metric with give us an idea of the complexity of our dataset (more genes detected    per UMI, more complex our data)
 - mitochondrial ratio: this metric will give us a percentage of cell reads originating from the mitochondrial genes

### Number of genes detected
This value is quite easy to calculate, as we take the log10 of the number of genes detected per cell and the log10 of the number of UMIs per cell, then divide the log10 number of genes by the log10 number of UMIs. per UMI
```{r}
# Add number of genes per UMI for each cell to metadata
merged_seurat$log10GenesPerUMI <- log10(merged_seurat$nFeature_RNA) / log10(merged_seurat$nCount_RNA)
```

### Mitochondrial ratio
Seurat has a convenient function that allows us to calculate the proportion of transcripts mapping to mitochondrial genes. The PercentageFeatureSet() function takes in a pattern argument and searches through all gene identifiers in the dataset for that pattern. Since we are looking for mitochondrial genes, we are searching any gene identifiers that begin with the pattern “MT-“. For each cell, the function takes the sum of counts across all genes (features) belonging to the “Mt-“ set, and then divides by the count sum for all genes (features). This value is multiplied by 100 to obtain a percentage value.

```{r}
# Compute percent mito ratio
merged_seurat$mitoRatio <- PercentageFeatureSet(object = merged_seurat, pattern = "^MT-")
merged_seurat$mitoRatio <- merged_seurat@meta.data$mitoRatio / 100
```

### Additional metadata columns
We are a now all set with quality metrics required for assessing our data. However, we would like to include some additional information that would be useful to have in our metadata including cell IDs and condition information.

When we added columns of information to our metadata file above, we simply added it directly to the metadata slot in the Seurat object using the $ operator. We could continue to do so for the next few columns of data, but instead we will extract the dataframe into a separate variable. In this way we can work with the metadata data frame as a seperate entity from the seurat object without the risk of affecting any other data stored inside the object.

Let’s begin by creating the metadata dataframe by extracting the meta.data slot from the Seurat object:
```{r}
# Create metadata dataframe
metadata <- merged_seurat@meta.data
```

Next, we’ll add a new column for cell identifiers. This information is currently located in the row names of our metadata dataframe. We will keep the rownames as is and duplicate it into a new column called cells:
```{r}
# Add cell IDs to metadata
metadata$cells <- rownames(metadata)
```

You should see that each cell ID has a ana_ or fuso_ prefix as we had specified when we merged the Seurat objects. We can use this prefix to create a new column indicating which condition each cell is classfied under. We will call this column sample:

```{r}
# Create sample column
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^ana1_"))] <- "ana1"
metadata$sample[which(str_detect(metadata$cells, "^ana2_"))] <- "ana2"
metadata$sample[which(str_detect(metadata$cells, "^ana3_"))] <- "ana3"
metadata$sample[which(str_detect(metadata$cells, "^ana4_"))] <- "ana4"
metadata$sample[which(str_detect(metadata$cells, "^ana5_"))] <- "ana5"
metadata$sample[which(str_detect(metadata$cells, "^ana6_"))] <- "ana6"
metadata$sample[which(str_detect(metadata$cells, "^fuso1_"))] <- "fuso1"
metadata$sample[which(str_detect(metadata$cells, "^fuso2_"))] <- "fuso2"
metadata$sample[which(str_detect(metadata$cells, "^fuso3_"))] <- "fuso3"

# Create condition
metadata$condition <- NA
metadata$condition[which(str_detect(metadata$cells, "^ana"))] <- "ana"
metadata$condition[which(str_detect(metadata$cells, "^fuso"))] <- "fuso"
```

And finally, we will rename some of the existing columns in our metadata dataframe to be more intuitive:

```{r}
# Rename columns
metadata <- metadata %>%
        dplyr::rename(seq_folder = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA)
```

Before we assess our metrics we are going to save all of the work we have done thus far back into our Seurat object. We can do this by simply assigning the dataframe into the meta.data slot:

```{r}
# Add metadata back to Seurat object
merged_seurat@meta.data <- metadata
```


```{r}
# Create .RData object to load at any time
save(merged_seurat, file="../data/seurat/merged_seurat.RData")
```



   
   
   
## Assessing the quality metrics
Now that we have generated the various metrics to assess, we can explore them with visualizations. We will assess various metrics and then decide on which cells are low quality and should be removed from the analysis:
 - Cell counts
 - UMI counts per cell
 - Genes detected per cell
 - UMIs vs. genes detected
 - Mitochondrial counts ratio
 - Novelty

### Cell counts
```{r}
# Visualize the number of cell counts per sample
metadata %>% 
  	ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")
```

### UMI counts (transcripts) per cell
The UMI counts per cell should generally be above 500, that is the low end of what we expect. If UMI counts are between 500-1000 counts, it is usable but the cells probably should have been sequenced more deeply.
```{r}
# Visualize the number UMIs/transcripts per cell
metadata %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)
```

### Genes detected per cell
We have similar expectations for gene detection as for UMI detection, although it may be a bit lower than UMIs. For high quality data, the proportional histogram should contain a single large peak that represents cells that were encapsulated. If we see a small shoulder to the left of the major peak, or a bimodal distribution of the cells, that can indicate a couple of things. It might be that there are a set of cells that failed for some reason. It could also be that there are biologically different types of cells (i.e. quiescent cell populations, less complex cells of interest), and/or one type is much smaller than the other (i.e. cells with high counts may be cells that are larger in size). Therefore, this threshold should be assessed with other metrics that we describe in this lesson.

```{r}
# Visualize the distribution of genes detected per cell via histogram
metadata %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)

# Visualize the distribution of genes detected per cell via boxplot
metadata %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells vs NGenes")
```

```{r}

metadata %>%
   group_by(condition) %>%
   summarise_at(vars(nGene), list(name = mean))
```

### UMIs vs genes detected
Two metrics that are often evaluated together are the number of UMIs and the number of genes detected per cell. Here, we have plotted the number of genes versus the number of UMIs coloured by the fraction of mitochondrial reads. Mitochondrial read fractions are only high in particularly low count cells with few detected genes (darker colored data points). This could be indicative of damaged/dying cells whose cytoplasmic mRNA has leaked out through a broken membrane, and thus, only mRNA located in the mitochondria is still conserved. These cells are filtered out by our count and gene number thresholds. Jointly visualizing the count and gene thresholds shows the joint filtering effect.

Cells that are poor quality are likely to have low genes and UMIs per cell, and correspond to the data points in the bottom left quadrant of the plot. Good cells will generally exhibit both higher number of genes per cell and higher numbers of UMIs.

With this plot we also evaluate the slope of the line, and any scatter of data points in the bottom right hand quadrant of the plot. These cells have a high number of UMIs but only a few number of genes. These could be dying cells, but also could represent a population of a low complexity celltype (i.e red blood cells).

```{r}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample)
```


### Mitochondrial counts ratio
This metric can identify whether there is a large amount of mitochondrial contamination from dead or dying cells. We define poor quality samples for mitochondrial counts as cells which surpass the 0.2 mitochondrial ratio mark, unless of course you are expecting this in your sample.

```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)
```


### Complexity 
We can evaluate each cell in terms of how complex the RNA species are by using a measure called the novelty score. The novelty score is computed by taking the ratio of nGenes over nUMI. If there are many captured transcripts (high nUMI) and a low number of genes detected in a cell, this likely means that you only captured a low number of genes and simply sequenced transcripts from those lower number of genes over and over again. These low complexity (low novelty) cells could represent a specific cell type (i.e. red blood cells which lack a typical transcriptome), or could be due to some other strange artifact or contamination. Generally, we expect the novelty score to be above 0.80 for good quality cells.

```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)
```

## Filtering
### Cell-level filtering
Now that we have visualized the various metrics, we can decide on the thresholds to apply which will result in the removal of low quality cells. Often the recommendations mentioned earlier are a rough guideline, and the specific experiment needs to inform the exact thresholds chosen. We will use the following thresholds:
 - nUMI > 500
 - nGene > 250
 - log10GenesPerUMI > 0.8
 - mitoRatio < 0.25

To filter, we wil go back to our Seurat object and use the subset() function:
```{r}
# Filter out low quality cells using selected thresholds - these will change with experiment
filtered_seurat <- subset(x = merged_seurat, 
                         subset= (nUMI >= 500) & 
                           (nGene >= 250) & 
                           (log10GenesPerUMI > 0.80) & 
                           (mitoRatio < 0.25))
			                          
```

### Gene-Level filtering
Within our data we will have many genes with zero counts. These genes can dramatically reduce the average expression for a cell and so we will remove them from our data. We will start by identifying which genes have a zero count in each cell:
```{r}
# Extract counts
counts <- GetAssayData(object = filtered_seurat, slot = "counts")

# Output a logical matrix specifying for each gene on whether or not there are more than zero counts per cell
nonzero <- counts > 0
```

Now, we will perform some filtering by prevalence. If a gene is only expressed in a handful of cells, it is not particularly meaningful as it still brings down the averages for all other cells it is not expressed in. For our data we choose to keep only genes which are expressed in 10 or more cells. By using this filter, genes which have zero counts in all cells will effectively be removed.

```{r}
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]
```
Finally, take those filtered counts and create a new Seurat object for downstream analysis.

```{r}
# Reassign to filtered Seurat object
filtered_seurat <- CreateSeuratObject(filtered_counts, meta.data = filtered_seurat@meta.data)
```

## Re-assess QC metrics
After performing the filtering, it’s recommended to look back over the metrics to make sure that your data matches your expectations and is good for downstream analysis.

```{r}
# Save filtered subset to new metadata
 metadata_clean <- filtered_seurat@meta.data
```

### Cell counts
```{r}
# Visualize the number of cell counts per sample
metadata_clean %>% 
  	ggplot(aes(x=sample, fill=sample)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")
```

### UMI counts (transcripts) per cell
```{r}
# Visualize the number UMIs/transcripts per cell
metadata_clean %>% 
  	ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)
```

### Genes detected per cell
```{r}
# Visualize the distribution of genes detected per cell via histogram
metadata_clean %>% 
  	ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)

# Visualize the distribution of genes detected per cell via boxplot
metadata_clean %>% 
  	ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells vs NGenes")
```
```{r}
metadata_clean %>%
   group_by(sample) %>%
   summarise_at(vars(nGene), list(name = mean))
```

```{r}
metadata_clean %>%
   group_by(condition) %>%
   summarise_at(vars(nGene), list(name = mean))
```

### UMIs vs genes detected
```{r}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata_clean %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~sample)
```

### Mitochondrial counts ratio
```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata_clean %>% 
  	ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)
```

### Complexity 
```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
metadata_clean %>%
  	ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)
```

# Save filtered cell data
```{r}
# Create .rds object to load at any time
saveRDS(filtered_seurat, file="../data/seurat/seurat_filtered.rds")
```


